File: 3/zad1/zad1.4/src/main/java/hr/fer/zemris/ooup/lab3/zad1/model/plugins/Parrot.java
package hr.fer.zemris.ooup.lab3.zad1.model.plugins;

import hr.fer.zemris.ooup.lab3.zad1.model.Animal;

public class Parrot extends Animal {

    private final String name;

    public Parrot(String name) {
        this.name = name;
    }

    @Override
    public String name() {
        return name;
    }

    @Override
    public String greet() {
        return "grr!";
    }

    @Override
    public String menu() {
        return "sjemenke";
    }

}

File: 3/zad1/zad1.1/parrot.c
#include <stdlib.h> 

typedef char const* (*PTRFUN)();

typedef struct {
	PTRFUN* vtable;
	char const* name;
} Parrot;

char const* name(Parrot* p) { return p->name; }
char const* greet(void) { return "grr!"; }
char const* menu(void) { return "sjemenke"; }

PTRFUN vtable[] = {(PTRFUN)name, greet, menu};

void* create(char const* name) {
	Parrot* p = (Parrot*)malloc(sizeof(Parrot));
	p->vtable = vtable;
	p->name = name;
	return p;
}

void* createStack(char const* name) {
	Parrot p;
	p.vtable = vtable;
	p.name = name;
	return &p;
}
